crates/uv-pep440/src/version.rs

```diff
@@ -933,14 +934,13 @@ impl FromStr for Version {
 struct VersionSmall {
     /// The number of segments in the release component.
     ///
-    /// Strictly speaking, this isn't necessary since `1.2` is considered
-    /// equivalent to `1.2.0.0`. But in practice it's nice to be able
-    /// to truncate the zero components. And always filling out to 4
-    /// places somewhat exposes internal details, since the "full" version
-    /// representation would not do that.
+    /// PEP 440 considers `1.2`  equivalent to `1.2.0.0`, but we want to preserve trailing zeroes
+    /// in roundtrips, as the "full" version representation also does.
     len: u8,
     /// The representation discussed above.
     repr: u64,
+    /// Force a niche into the aligned type so the [`Version`] enum is two words instead of three.
+    _force_niche: NonZero<u8>,
```

Comment:
This is interesting. I get it. Even this only gives us one niche value (which I guess is all we need here), but I believe rustc itself is capable of expressing a _range_ of niche values: https://doc.rust-lang.org/src/core/num/nonzero.rs.html#50

Would be nice to have access to that more directly... Some day hopefully.

Reply:
We have a bunch of enum types (#10389 too) that could shrink a lot by rustc understanding niches across type nesting trees.